### ISSUE TITLE
Creating a file and directory structure

### ISSUE THREAD

**Issue by budmil** [privML/privacy-evaluator#5]
### Acceptance criteria
modular codebase
modular attack classes
attacks should have roughly the same interface (create actual interface that attack classes would implement, with methods like: test, outputMetric)
accept attack parameters (for ART)

**Comment by Erdnaf** [comment]
### First idea of structuring
**/main.py -** main script that is called with a path to a modal and the specified attack plus parameters for that attack for testing.
**/models/ -** in this folder we store our pretrained models for testing
**/modules/ -** in this folder we store the attack wrappers all ending with -attack.py and implementing the attack interface class

For this structure we would import and not fork the adversarial-robustness-toolbox.
Also the model training would not be in the repository

Feedback apreciated

**Comment by blauertee** [comment]
Hey this would make sense but i think it doe's not quite agree with python structuring [conventions](https://docs.python-guide.org/writing/structure/). ([here](https://github.com/ytdl-org/youtube-dl) is also an example of a bigger project, which uses as similar structure)

Acording to that I think we should do something like:

``
README.rst
LICENSE
requirements.txt
privacy-evaluator/__main__.py
privacy-evaluator/__init__.py
...
privacy-evaluator/attacks/membershipinference.py
...
privacy-evaluator/modules/metric.py
tests/test_basic.py
tests/test_advanced.py
``

I think it makes also sense to seperate attack modules from components that directly belong to our software, since these should be easy to add and manipulate by users.

Feedback still appreciated!

**Comment by blauertee** [comment]
I think according to conventions we'd have to place single file modules not in the modules, but in the `privacy-evaluator` folder. Now I feel like I wanna violate conventions ... what do you think?

**Comment by marisanest** [comment]
Hey, just a thought about the code conventions: Maybe you already have this in mind but it might be also helpful to define how code documentation should be done (e.g. in a separate doc directory as in [scikit-learn/doc](https://github.com/scikit-learn/scikit-learn/tree/main/doc) or directly within the code or both). In this context, you could also define code documentation conventions such as Sphinx Style or Google Style as in [docstrings-python](https://www.datacamp.com/community/tutorials/docstrings-python).

**Comment by Erdnaf** [comment]
We started to structure the interface here, feedback appreciated! 

https://drive.google.com/file/d/19oWHQ0ZkuA3BTzlUxaHVbBbFVn7vGJbm/view?usp=sharing

![Code Base PrivML(2)](https://user-images.githubusercontent.com/64256710/116098199-27115d80-a6ab-11eb-8c17-b3884d95a05f.png)

**Comment by blauertee** [comment]
This issue now focuses on file and directory structure, for interface and class structure discussion I opened #17.

**Comment by adrinjalali** [comment]
Regarding file names and modules, bare in mind that users can import from your modules as well as files. So if you have a file such as `attacks/example.py`, they can do `from privacy_evaluator.attacks.example import blahblah`. A convention is that everything starting with an `_` is _private_. This means you probably want your file names in your module folders to start with an `_`. Having single files in the root folder makes it quite hard for you to mask the private stuff and only expose the public part of your API.

Regarding tests, you can either have it as a separate folder with a folder structure which resembles and copies the folder structure under `privacy-evaluator`, or you could have a `tests` folder in each of your folders beside the modules and have tests there.

Regarding documentation, it depends on what kind of documentation you're talking about. The API docs are really nice if they're in the docstrings right where you write the code, and it's easy to keep them synced. There are disadvantages, but the pros outweigh the cons IMO. Then there are user guidelines and longer examples which should go into their own dedicated folders. User guidelines are usual `.rst` files where you can also include code, but they're mostly about the text and telling people how to use the software. Examples are `.py` files where you write something which can be converted into a document, as well as a python notebook for users to download and explore. You can use `sphinx-gallery` to convert the examples to notebooks and documents. For a simpler implementation you could also look at https://github.com/fairlearn/fairlearn repo.

**Comment by blauertee** [comment]
Thanks for the feedback!

> Having single files in the root folder makes it quite hard for you to mask the private stuff and only expose the public part of your API.

By root folder you mean the privacy_evaluator folder right and not the repos root folder right? So you suggest using a modules folder for private modules (e.g. `privacy_evaluator/modules/_priv_module/__init__.py` or `privacy_evaluator/modules/_priv_module.py`)   and for public importable modules we should have them right in the `privacy_evaluator` folder right?

**Comment by adrinjalali** [comment]
> By root folder you mean the privacy_evaluator folder right and not the repos root folder right?

Yes.

> So you suggest using a modules folder for private modules (e.g. `privacy_evaluator/modules/_priv_module/__init__.py` or `privacy_evaluator/modules/_priv_module.py`)   and for public importable modules we should have them right in the `privacy_evaluator` folder right?

Yeah, I'd have it for instance as `privacy_evaluator/attacks/_attack1.py`, and inside `privacy_evaluator/attacks/__init__.py` I'd have:

``` python
from .attack1 import AttackClass
...

__all__ = [
    "AttackClass",
    ...
]
```

And then user code would look like

``` python
from privacy_evaluator.attacks import AttackClass
```
